### ISSUE TITLE
HTTP transport for Thrift

### ISSUE THREAD

**Issue by ii64** [cloudwego/volo#125]
## Feature Request

Support for HTTP server/client transport wrapper for Thrift RPC.

### Crates

volo-thrift

### Motivation

Extending compatibility of Thrift implementation to support HTTP transport wrapper.
Current state of supported features compared with another Thrift library: [apache/thrift/LANGUAGES.md](https://github.com/apache/thrift/blob/master/LANGUAGES.md)

### Proposal

Using `hyper` as Thrift HTTP transport wrapper,  implements `volo::net::dial::MakeTransport`.

### Alternatives

RFC

**Comment by PureWhiteWu** [comment]
Is there a spec that describes the HTTP transport?
I didn't find it.

**Comment by ii64** [comment]
Me neither, people were able to implement them in any language and do cross-test between each of the implementation so I think there should be a general idea. I agree that we need a spec about that. Apache Thrift had it supported since 0.1.x https://github.com/apache/thrift/blob/0.1.x/lib/cpp/src/transport/THttpClient.cpp

**Comment by PureWhiteWu** [comment]
Maybe we should submit an issue for that to the Apache thrift repo?

**Comment by ii64** [comment]
apache/thrift repository seem don't accept issue from GitHub, Apache's Jira public signup is also disabled (https://infra.apache.org/jira-guidelines.html#who).

Theoretically speaking, can HTTP Transport implemented with only current `impl MakeTransport`? 
I've looked into this a bit, `impl MakeTransport` has:
```rust
async fn make_transport(&self,addr:volo::net::Address) -> std::io::Result<(Self::ReadHalf,Self::WriteHalf)>
```
And think that it's not the correct way of open a connection (over HTTP) since it need to be `AsyncWrite` and `AsyncRead` at the same time, `AsyncWrite` has `poll_flush` but it can't be implemented in there since there will be a new HTTP response `Body` to be consumed by the `Decoder`.

This may need to implement another Client beside existing multiplex and pingpong, the `motore::{Service,UnaryService}`. What do you think about this?

**Comment by thaodt** [comment]
hi, do we still need it? If yes, I want to give it a try

**Comment by ii64** [comment]
Absolutely, that would be fantastic! We can figure out later on how we are going to merge the feature. Your offer of help is greatly appreciated. Thanks

**Comment by ii64** [comment]
Hi, please kindly check the PR.

Currently the implementation is using `reqwest` and only client. I'd like to see this stabilized on `volo` and using our abstraction on `volo-http` but aside from that, it is pretty much functioning and I wonder if we can roll-out that as we've proposed the public API related for HTTP transport.

E.g.

```rust
let url = "https://127.0.0.1/thrift-http-endpoint".parse::<hyper::Uri>().unwrap();
let url = volo::net::Address::from(url);
let client = volo_thrift::client::ClientBuilder::
        new("dummy", dummys::MkDummyServiceGenericClient)
        .make_codec(mk_codec)
        .address(url)
        .header("user-agent", "volo-http/1.0".parse().unwrap())
        .build();
```

More detailed example can be accessed on https://github.com/ii64/volo_testing/blob/f796b7360e32493b450f907d27a048c936261b1d/the_example/src/bin/client_http_testing.rs

**Comment by thaodt** [comment]
@ii64 at a first glance, it seems great. Sorry for lack of activity for a long time. But im curious on your implementation, I will check your PR also.

**Comment by ii64** [comment]
No worries, we are still doing RFC, you can check the discussion on linked PR
